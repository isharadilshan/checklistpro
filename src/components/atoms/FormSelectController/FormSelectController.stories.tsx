import React from 'react';
import {NativeBaseProvider} from 'native-base';
import {useForm} from 'react-hook-form';
import {ComponentMeta, ComponentStory} from '@storybook/react';

import FormSelectController from './';

export default {
  title: 'components/atoms/FormSelectController',
  component: FormSelectController,
} as ComponentMeta<typeof FormSelectController>;

export const Basic: ComponentStory<typeof FormSelectController> = () => {
  const {control} = useForm();

  const args = {
    name: 'Form Input' || undefined,
    defaultValue: 'Default Value',
    error: 'ERROR',
    type: 'text',
    control: control,
    items: [
      {key: '1', label: 'Food', value: 'FOOD'},
      {key: '2', label: 'Medical', value: 'MEDICAL'},
      {key: '3', label: 'Transport', value: 'TRANSPORT'},
    ],
  };
  return (
    <NativeBaseProvider>
      <FormSelectController {...args} />
    </NativeBaseProvider>
  );
};
