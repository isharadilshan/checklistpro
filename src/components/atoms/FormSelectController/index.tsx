import React from 'react';
import {FormControl} from 'native-base';
import {
  Control,
  Controller,
  FieldError,
  Path,
  RegisterOptions,
} from 'react-hook-form';
import ModalSelector from 'react-native-modal-selector';
import Icon from 'react-native-vector-icons/FontAwesome';

type SelectItem = {
  key: string;
  label: string;
  value: string;
};

interface FormSelectControllerProps<FieldsType> {
  name: Path<FieldsType>;
  defaultValue?: string;
  rules?: RegisterOptions;
  error?: FieldError;
  control: Control<any>;
  items?: Array<SelectItem>;
}

interface Props<FieldsType> extends FormSelectControllerProps<FieldsType> {
  label?: string;
  placeholder?: string;
}

const FormSelectController = <FieldsType,>({
  error,
  rules,
  label,
  control,
  name,
  placeholder,
  items = [],
}: Props<FieldsType>) => {
  const isInvalid = error != null;
  const isRequired = rules != null && 'required' in rules;

  return (
    <FormControl
      isInvalid={isInvalid}
      isRequired={isRequired}
      style={{marginBottom: 10}}
    >
      {label != null && <FormControl.Label>{label}</FormControl.Label>}
      <Controller
        control={control}
        name={name}
        rules={rules}
        render={({field: {onChange, value}}) => (
          <ModalSelector
            data={items}
            initValue={value}
            accessible={true}
            scrollViewAccessibilityLabel={'Scrollable options'}
            cancelButtonAccessibilityLabel={'Cancel Button'}
            onChange={(option) => onChange(option.label)}
          />
        )}
      />
      {error != null && (
        <FormControl.ErrorMessage>{error.message}</FormControl.ErrorMessage>
      )}
    </FormControl>
  );
};

export default FormSelectController;
