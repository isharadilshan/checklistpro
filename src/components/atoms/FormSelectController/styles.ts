import {ViewStyle, StyleSheet} from 'react-native';

interface Style {
  controlWrapper: ViewStyle;
}

export default () => {
  return StyleSheet.create<Style>({
    controlWrapper: {
      marginBottom: 10,
    },
  });
};
