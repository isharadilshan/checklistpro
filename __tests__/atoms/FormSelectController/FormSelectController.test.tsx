import React from 'react';
import {describe, expect} from '@jest/globals';
import {shallow, configure} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import FormSelectController from '../../../src/components/atoms/FormSelectController';

configure({adapter: new Adapter()});

const props = {
  name: 'TEST FORM SELECT CONTROLLER',
  defaultValue: 'test',
  error: 'This is not a valid form input',
  label: 'FORM SELECT',
  items: [
    {
      key: 1,
      label: 'ToDo',
      value: 'TODO',
    },
    {
      key: 2,
      label: 'In-Progress',
      value: 'INPROGRESS',
    },
    {
      key: 1,
      label: 'Hold',
      value: 'HOLD',
    },
    {
      key: 2,
      label: 'Done',
      value: 'DONE',
    },
  ],
};

describe('when rendering', () => {
  const wrapper = shallow(<FormSelectController {...props} />);

  it('should render a component matching the snapshot', () => {
    const tree = toJson(wrapper);
    expect(tree).toMatchSnapshot();
    expect(wrapper).toHaveLength(1);
  });
});
