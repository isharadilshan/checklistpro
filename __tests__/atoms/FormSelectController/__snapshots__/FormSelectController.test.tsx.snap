// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when rendering should render a component matching the snapshot 1`] = `
<Memo(ForwardRef(FormControl))
  isInvalid={true}
  isRequired={false}
  style={
    {
      "marginBottom": 10,
    }
  }
>
  <Memo(ForwardRef(FormControlLabel))>
    FORM SELECT
  </Memo(ForwardRef(FormControlLabel))>
  <Component
    name="TEST FORM SELECT CONTROLLER"
    render={[Function]}
  />
  <Memo(ForwardRef(FormControlErrorMessage)) />
</Memo(ForwardRef(FormControl))>
`;
